### ISSUE TITLE
Graceful interrupt not graceful

### ISSUE THREAD

**Issue by ruotianluo** [PyTorchLightning/pytorch-lightning#4111]
The current key interrupt will call on_train_end which will save the checkpoint. However when resuming from the saved checkpoint, it starts a new epoch, which is not graceful: presumably it should restart from the middle of the epoch(ideally); otherwise it should not save the interrupted checkpoint with the same name as normal checkpoints (for example, save as a new name).

**Comment by williamFalcon** [comment]
the training data are random... so, starting from the beginning or the middle or end don't make a difference.

**Comment by ruotianluo** [comment]
It matters for research where the number of epochs or the number of iterations should be strictly controlled for fair comparison.

**Comment by williamFalcon** [comment]
i know haha... we do research as well. but you set shuffle=True on the training data. And i don't think there is a way to restore a shuffled dataset to exactly the same spot.

**Comment by ruotianluo** [comment]
Sure, so my proposal is maybe handling keyinterrupt a little bit different; for example, at least save the checkpoint with a different name.
(There is a way to restore shuffled dataset when using ddp, because the distributed sampler is somewhat deterministic. In other cases, there are other nasty ways.)

**Comment by williamFalcon** [comment]
got it. yeah, that would be an awesome contribution if you want to take a stab at it! we’ll help you complete it.

**Comment by ruotianluo** [comment]
To whom gets to this issue, I have a solution that requires a minimum change in training code.
https://github.com/ruotianluo/rtutils#stateless-resume-from-checkpoint-saved-in-the-middle-of-one-epoch
